### PR TITLE
Notification inline actions - preliminary refactor

### PR DESCRIPTION
**Pull request by mkevins**
### Description

This PR is a preliminary step to refactor and align shared elements of the design for inline actions on notifications. In the development of resolving https://github.com/wordpress-mobile/WordPress-Android/issues/20040, a strategy emerged that would benefit other similar issues / tasks: establishing a declarative model for `Notification`s, and treating access to the pre-existing class(es) as an implementation detail. This allows us to treat the existing code structure itself as "immutable", a benefit since any changes there may be hard to follow in the current state.

This PR also adds many missing nullability annotations, and extracts _some_ logic from a method which violates lint for `"CyclomaticComplexMethod"` and `"LongMethod"` rules (though more work is needed before the rule suppression can be removed).

-----

## To Test:

* Run the app using this branch (either build in Android Studio, or via one of the apks)
* Login to the app using an account which has some activity (with various kinds of notifications)
* Smoke test the notifications behavior:
  * Scroll the notifications list and ensure they all look correct
  * Tap on some of the notifications to ensure that the behavior is still correct (e.g. tap a like notification to see the list of users who liked the comment or post, etc.)
* Observe that comment notifications have a star icon
* Observe that like notifications have a share icon
  * Please note that the share icon is just a place-holder for now, and currently does not match the icon from the Figma designs

-----

## Regression Notes

1. Potential unintended areas of impact

Notifications

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manual testing

3. What automated tests I added (or what prevented me from doing so)

Additional refactoring is necessary before automated tests can be added

-----

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## Testing Checklist:

- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>Notification</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>Like</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>NoteType</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by jarvislin — APPROVED** [review]
Thanks for refactoring the code, this is helpful to me!
Code looks great, and the testing result is as expected.

**Comment by mkevins** [comment]
Thank you for reviewing and testing!

**Review by antonis — APPROVED** [review]
Thank you for refactoring this part of the code @mkevins 🙇 
The code looks good and I haven't noticed any behavior changes in my tests 🎉
